client/build.go

```diff
@@ -108,3 +108,18 @@ func (g *gatewayClientForBuild) Inputs(ctx context.Context, in *gatewayapi.Input
 	ctx = buildid.AppendToOutgoingContext(ctx, g.buildID)
 	return g.gateway.Inputs(ctx, in, opts...)
 }
+
+func (g *gatewayClientForBuild) NewContainer(ctx context.Context, in *gatewayapi.NewContainerRequest, opts ...grpc.CallOption) (*gatewayapi.NewContainerResponse, error) {
+	ctx = buildid.AppendToOutgoingContext(ctx, g.buildID)
```

Comment:
I wonder if these should check for cap and return right away. I see previous ones don't but could give a better error than "method not supported"

Reply:
This seems doable, I just wanted to double check that I am on the right path here.  To get this done we will have to add to the `caps apicaps.CapSet` to the `gatewayClientForBuild` struct, and initialize it with a `Ping` call in the `(*Client) gatewayClientForBuild` call.  This would effectively be the same implementation currently in `grpcclient.New`.  It seems we would also have to move the `defaultCaps` to be public in a common package (seems like we could move the default functions to the `frontend.gateway.pb` package?). Then we can add the `c.caps.Supports(...)` calls to most of the `(*gatewayClientForBuild)` functions.  Does this sound right?

Reply:
We already do the `Ping` call and `defaultCaps` check in `client.New()` so I think the only thing you need to do is add a client property for `gatewayClientForBuild` and set it after calling `client.New()`.

Reply:
I am probably missing something, not sure this is what you had in mind:
https://github.com/moby/buildkit/pull/1627/commits/2f7ce818e6d8c321d29a9c6961c2eefd997ff693

Reply:
Pretty much, except you don't need to extend client with `Caps()`, you can just do `c.BuildOpts().Caps`

Reply:
I see, I was getting confused since we don't have an explicit `gateway.client.Client` in that code, I wasn't thinking about the `*grpcclient` implementing both the `GrpcClient` and `gateway.client.Client` interfaces, but a cast was needed.  

[Updated](
https://github.com/moby/buildkit/pull/1627/commits/efb36aa53c5399c3041bdf0aded8c045ab9d2f53#diff-93c3f24a34a25d25d12623f602db4974R55)